File: btypes_big_integer/src/main/rust/btypes/src/binteger.rs
use crate::bboolean::BBoolean;
use crate::bobject::BObject;

use num_bigint::{BigInt, BigUint};
use num_traits::pow::Pow;
use core::ops::Rem;
use std::fmt;
use std::hash::Hash;
use std::ops::{Neg};

pub trait BInt: BObject { fn get_binteger_value(&self) -> BInteger { panic!("get_integer_value not implemented!"); }}
pub trait FromBInt { fn from<T: BInt>(value: &T) -> Self; }

#[derive(Default, Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub struct BInteger {
    val: BigInt,
}

impl BObject for BInteger {}
impl BInt for BInteger {
    fn get_binteger_value(&self) -> BInteger { return self.clone(); }
}

impl FromBInt for BInteger {
    fn from<T: BInt>(value: &T) -> Self {
        return value.get_binteger_value();
    }
}

impl BInteger {
    #![allow(non_snake_case, dead_code)]

    pub fn new(init: &[u8]) -> BInteger {
        return BInteger {
            val: BigInt::parse_bytes(init, 10).unwrap(),
        }
    }

    pub const fn new_from_bigint(init: BigInt) -> BInteger {
        return BInteger {
            val: init,
        }
    }

    pub fn get_val(&self) -> BigInt { return self.val.clone(); }

    pub fn lessEqual(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val <= o.val);
    }

    pub fn greaterEqual(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val >= o.val);
    }

    pub fn less(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val < o.val);
    }

    pub fn greater(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val > o.val);
    }

    pub fn equal(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val == o.val);
    }

    pub fn unequal(&self, o: &BInteger) -> BBoolean {
        return BBoolean::new(self.val != o.val);
    }

    pub fn plus(&self, v: &BInteger) -> BInteger {
        return BInteger::new_from_bigint(&self.val + v.val.clone());
    }

    pub fn minus(&self, v: &BInteger) -> BInteger {
        return BInteger::new_from_bigint(&self.val - v.val.clone());
    }

    pub fn multiply(&self, v: &BInteger) -> BInteger {
        return BInteger::new_from_bigint(&self.val * v.val.clone());
    }

    pub fn power(&self, v: &BInteger) -> BInteger {
        if v.val.sign() == num_bigint::Sign::Minus { panic!("Power with negative exponent!") }
        return BInteger::new_from_bigint(Pow::<BigUint>::pow(&self.val, v.val.to_biguint().unwrap()));
    }

    pub fn divide(&self, v: &BInteger) -> BInteger {
        return BInteger::new_from_bigint(&self.val / v.val.clone());
    }

    pub fn modulo(&self, v: &BInteger) -> BInteger {
        return BInteger::new_from_bigint(Rem::rem(&self.val, v.val.clone()));
    }

    pub fn succ(&self) -> BInteger {
        return BInteger::new_from_bigint(&self.val + BigInt::from(1));
    }

    pub fn inc(&mut self) {
        self.val = &self.val + BigInt::from(1);
    }

    pub fn pred(&self) -> BInteger {
        return BInteger::new_from_bigint(&self.val - BigInt::from(1));
    }

    pub fn negative(&self) -> BInteger {
        return BInteger::new_from_bigint((&self.val).neg());
    }

    pub fn positive(&self) -> BInteger {
        return self.clone();
    }

    pub fn isInteger() -> BBoolean {
        return BBoolean::new(true);
    }

    pub fn isNotInteger() -> BBoolean {
        return BBoolean::new(false);
    }

    pub fn isNatural(&self) -> BBoolean {
        return BBoolean::new(self.val.sign() != num_bigint::Sign::Minus);
    }

    pub fn isNotNatural(&self) -> BBoolean {
        return BBoolean::new(self.val.sign() == num_bigint::Sign::Minus);
    }

    pub fn isNatural1(&self) -> BBoolean {
        return BBoolean::new(self.val.sign() == num_bigint::Sign::Plus);
    }

    pub fn isNotNatural1(&self) -> BBoolean {
        return BBoolean::new(self.val.sign() != num_bigint::Sign::Plus);
    }
}

impl fmt::Display for BInteger {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.val)
    }
}

File: btypes_big_integer/src/main/js/btypes/BInteger.ts
import {BBoolean} from './BBoolean.js';
import {BObject} from "./BObject.js";

export class BInteger implements BObject{

    value: number;

    constructor(value: number) {
        this.value = value;
    }

    equals(obj: any): boolean {
        if (obj instanceof BInteger) {
            return this.compareTo(obj) === 0;
        }
        return false;
    }

    static build(value: number): BInteger {
        return new BInteger(value);
    }

    compareTo(o: BInteger): number {
        return this.value - o.value;
    }

    lessEqual(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) <= 0);
    }


    greaterEqual(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) >= 0);
    }

    less(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) < 0);
    }

    greater(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) > 0);
    }

    equal(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) === 0);
    }

    unequal(o: BInteger): BBoolean {
        return new BBoolean(this.compareTo(o) != 0);
    }
    
    intValue(): number {
        return this.value;
    }

    plus(o: BInteger): BInteger {
        return new BInteger(this.value + o.value);
    }

    toString(): string {
        return this.value.toString();
    }

    minus(o: BInteger): BInteger {
        return new BInteger(this.value - o.value);
    }

    multiply(o: BInteger): BInteger {
        return new BInteger(this.value * o.value);
    }

    power(o: BInteger): BInteger {
        return new BInteger(this.value ^ o.value);
    }

    divide(o: BInteger): BInteger {
        return new BInteger(~~(this.value / o.value));
    }

    modulo(o: BInteger): BInteger {
        return new BInteger(this.value % o.value);
    }

    succ(): BInteger  {
        return new BInteger(this.value + 1);
    }

    pred(): BInteger  {
        return new BInteger(this.value - 1);
    }

    leftShift(): BInteger  {
        return new BInteger(this.value << 1);
    }

    rightShift(): BInteger  {
        return new BInteger(this.value >> 1);
    }

    isCase(o: BInteger): boolean {
        return this.equals(o);
    }

    negative(): BInteger  {
        return new BInteger(-this.value);
    }

    positive(): BInteger  {
        return this;
    }

    getValue(): number {
        return this.value;
    }

    isInteger(): BBoolean {
        return new BBoolean(true);
    }

    isNotInteger(): BBoolean {
        return new BBoolean(false);
    }

    isNatural(): BBoolean {
        return this.greaterEqual(new BInteger(0));
    }

    isNotNatural(): BBoolean {
        return this.isNatural().not();
    }

    isNatural1(): BBoolean {
        return this.greater(new BInteger(0));
    }

    isNotNatural1(): BBoolean {
        return this.isNatural1().not();
    }

    hashCode(): number {
        const prime = 31;
        let result = 1;
        result = prime * result + this.value;
        return result;
    }
}